### PR TITLE
refactor(telemetry): clean up span creation count metrics

### PR DESCRIPTION
**Pull request by mabdinur**
Reduces code duplication but also decreases readability 😅 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](../docs/contributing.rst#release-branch-maintenance))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](../docs/contributing.rst#release-branch-maintenance)

### CODE REVIEW

File: ddtrace/internal/processor/trace.py [hunk 1]
```diff
@@ -258,6 +242,9 @@ def on_span_finish(self, span):
                     except Exception:
                         log.error("error applying processor %r", tp, exc_info=True)
 
+                # perf: telemetry_metrics_writer.add_count_metric(...) is an expensive operation.
+                # We should avoid calling this method on every invocation of ``SpanAggregator.on_span_finish()``
+                self._queue_span_count_metrics("span_finished", "integration_name")
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:247]:
```suggestion
                self._queue_span_count_metrics("spans_finished", "integration_name")
```

File: ddtrace/internal/processor/trace.py [hunk 2]
```diff
@@ -276,20 +263,10 @@ def shutdown(self, timeout):
...
+        self._queue_span_count_metrics("spans_finished", "integration_name", None)
         # on_span_finish(...) queues span finish metrics in batches of 100. This ensures all remaining counts are sent
         # before the tracer is shutdown.
-        for api, count in self._span_api_to_count["spans_finished"].items():
-            telemetry_metrics_writer.add_count_metric(
-                TELEMETRY_NAMESPACE_TAG_TRACER, "spans_finished", count, tags=(("integration_name", api),)
-            )
-        self._span_api_to_count["spans_finished"] = defaultdict(int)
-
+        self._queue_span_count_metrics("spans_created", "integration_name", None)
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:269]:
```suggestion
        self._queue_span_count_metrics("spans_finished", "integration_name", None)
```

File: ddtrace/internal/processor/trace.py [hunk 3]
```diff
@@ -197,20 +197,15 @@ def on_span_start(self, span):
...
-                    telemetry_metrics_writer.add_count_metric(
-                        TELEMETRY_NAMESPACE_TAG_TRACER, "spans_created", count, tags=(("integration_name", api),)
-                    )
-                self._span_api_to_count["spans_created"] = defaultdict(int)
+            self._queue_span_count_metrics("spans_created", "integration_name")
 
     def on_span_finish(self, span):
         # type: (Span) -> None
         with self._lock:
+            self._span_metrics["span_finished"][span._span_api] += 1
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:208]:
```suggestion
            self._span_metrics["spans_finished"][span._span_api] += 1
```

File: ddtrace/internal/processor/trace.py [hunk 4]
```diff
@@ -298,6 +275,17 @@ def shutdown(self, timeout):
...
 
+    def _queue_span_count_metrics(self, metric_name, tag_name, min_count=100):
+        # type: (str, str, Optional[int]) -> None
+        """Queues a telemetry count metric for span created and span finished"""
+        if min_count is None or sum(self._span_metrics[metric_name].values()) >= min_count:
+            for tag_value, count in self._span_metrics[metric_name].items():
+                telemetry_metrics_writer.add_count_metric(
+                    TELEMETRY_NAMESPACE_TAG_TRACER, metric_name, count, tags=((tag_name, tag_value),)
+                )
+                print("{} {}".format(tag_value, count))
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:286]:
```suggestion
```

**Review by ZStriker19 — DISMISSED** [review]

File: ddtrace/internal/processor/trace.py [hunk 5]
```diff
@@ -258,6 +242,9 @@ def on_span_finish(self, span):
                     except Exception:
                         log.error("error applying processor %r", tp, exc_info=True)
 
+                # perf: telemetry_metrics_writer.add_count_metric(...) is an expensive operation.
+                # We should avoid calling this method on every invocation of ``SpanAggregator.on_span_finish()``
```
Comment (Yun-Kim) [inline ddtrace/internal/processor/trace.py:246]:
```suggestion
```
Can we move this comment (and all instances of it) to the new method?

File: ddtrace/internal/processor/trace.py [hunk 6]
```diff
@@ -298,6 +275,16 @@ def shutdown(self, timeout):
             # It's possible the writer never got started in the first place :(
             pass
 
+    def _queue_span_count_metrics(self, metric_name, tag_name, min_count=100):
+        # type: (str, str, Optional[int]) -> None
+        """Queues a telemetry count metric for span created and span finished"""
+        if min_count is None or sum(self._span_metrics[metric_name].values()) >= min_count:
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:281]:
```suggestion
        # perf: telemetry_metrics_writer.add_count_metric(...) is an expensive operation.
        # We should avoid calling this method on every invocation of ``SpanAggregator.on_span_finish()``
        if min_count is None or sum(self._span_metrics[metric_name].values()) >= min_count:
```

File: ddtrace/internal/processor/trace.py [hunk 7]
```diff
@@ -197,20 +197,15 @@ def on_span_start(self, span):
         with self._lock:
             trace = self._traces[span.trace_id]
             trace.spans.append(span)
-            self._span_api_to_count["spans_created"][span._span_api] += 1
+            self._span_metrics["spans_created"][span._span_api] += 1
             # perf: telemetry_metrics_writer.add_count_metric(...) is an expensive operation.
             # We should avoid calling this method on every invocation of ``SpanAggregator.on_span_start()``
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:202]:
```suggestion
```

**Review by Yun-Kim — APPROVED** [review]

File: ddtrace/internal/processor/trace.py [hunk 8]
```diff
@@ -277,28 +260,31 @@ def shutdown(self, timeout):
...
             pass
 
+    def _queue_span_count_metrics(self, metric_name, tag_name, min_count=100):
+        # type: (str, str, Optional[int]) -> None
+        """Queues a telemetry count metric for span created and span finished"""
+        # perf: telemetry_metrics_writer.add_count_metric(...) is an expensive operation.
+        # We should avoid calling this method on every invocation of span finish and span start.
+        if min_count is None or sum(self._span_metrics[metric_name].values()) >= min_count:
+            for tag_value, count in self._span_metrics[metric_name].items():
+                telemetry_metrics_writer.add_count_metric(
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:283]:
```suggestion
                telemetry_writer.add_count_metric(
```

File: ddtrace/internal/processor/trace.py [hunk 9]
```diff
@@ -277,28 +260,31 @@ def shutdown(self, timeout):
...
-                telemetry_writer.add_count_metric(
-                    TELEMETRY_NAMESPACE_TAG_TRACER, "spans_created", count, tags=(("integration_name", api),)
-                )
-            self._span_api_to_count["spans_created"] = defaultdict(int)
+            self._queue_span_count_metrics("spans_created", "integration_name", None)
+            # on_span_finish(...) queues span finish metrics in batches of 100. This ensures all remaining counts are sent
+            # before the tracer is shutdown.
+            self._queue_span_count_metrics("spans_finished", "integration_name", None)
+            # The telemetry metrics writer can be shutdown before the tracer. This ensures all tracer metrics always sent.
+            telemetry_metrics_writer.periodic()
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:268]:
```suggestion
            telemetry_writer.periodic()
```

File: ddtrace/internal/processor/trace.py [hunk 10]
```diff
@@ -277,28 +260,31 @@ def shutdown(self, timeout):
...
             # on_span_start queue span created counts in batches of 100. This ensures all remaining counts are sent
             # before the tracer is shutdown.
-            for api, count in self._span_api_to_count["spans_created"].items():
-                telemetry_writer.add_count_metric(
-                    TELEMETRY_NAMESPACE_TAG_TRACER, "spans_created", count, tags=(("integration_name", api),)
-                )
-            self._span_api_to_count["spans_created"] = defaultdict(int)
+            self._queue_span_count_metrics("spans_created", "integration_name", None)
+            # on_span_finish(...) queues span finish metrics in batches of 100. This ensures all remaining counts are sent
+            # before the tracer is shutdown.
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:265]:
```suggestion
            # on_span_finish(...) queues span finish metrics in batches of 100. 
            # This ensures all remaining counts are sent before the tracer is shutdown.
```

File: ddtrace/internal/processor/trace.py [hunk 11]
```diff
@@ -277,28 +260,31 @@ def shutdown(self, timeout):
...
+            self._queue_span_count_metrics("spans_finished", "integration_name", None)
+            # The telemetry metrics writer can be shutdown before the tracer. This ensures all tracer metrics always sent.
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:267]:
```suggestion
            # The telemetry metrics writer can be shutdown before the tracer.
            # This ensures all tracer metrics always sent.
```

File: ddtrace/internal/processor/trace.py [hunk 12]
```diff
@@ -277,28 +260,32 @@ def shutdown(self, timeout):
         if self._span_api_to_count["spans_created"] or self._span_api_to_count["spans_finished"]:
```
Comment (mabdinur) [inline ddtrace/internal/processor/trace.py:260]:
```suggestion
        if self._span_metrics["spans_created"] or self._span_metrics["spans_finished"]:
```

**Review by Yun-Kim — APPROVED** [review]

**Pull request merged**